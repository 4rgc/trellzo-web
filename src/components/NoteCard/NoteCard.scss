.card.card-noimage {
	align-items: start;
	justify-content: flex-start;

	.card-title {
		margin-top: 0;
		text-align: left;
	}
	.card-content-text {
		text-align: left;
		font-size: small;
	}
}
