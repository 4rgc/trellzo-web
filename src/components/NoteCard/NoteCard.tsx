import Card from '../Card';
import PartialNote from '../../types/PartialNote';
import './NoteCard.scss';
import { DateTime, Duration } from 'luxon';
import humanizeDuration from 'humanize-duration';

export type NoteCardProps = {
	note: PartialNote;
};

const stripTime = (d: DateTime) => {
	return Duration.fromDurationLike({
		year: d.year,
		month: d.month,
		day: d.day,
	});
};

const NoteCardContent: React.FC<{
	description: string;
	startDate?: string;
	dueDate?: string;
}> = ({ description, dueDate }) => {
	let dueComponent;

	if (dueDate) {
		const daysUntilDue =
			stripTime(DateTime.fromISO(dueDate)).days -
			stripTime(DateTime.now()).days;
		const msecUntilDue = Duration.fromDurationLike({
			days: daysUntilDue,
		}).toMillis();

		const humanizedDaysUntilDue = humanizeDuration(msecUntilDue, {
			units: ['d'],
			round: true,
		});

		if (daysUntilDue > 3) {
			dueComponent = (
				<>
					due in{' '}
					<span className="due-far">{humanizedDaysUntilDue}</span>
				</>
			);
		} else if (daysUntilDue > 1) {
			dueComponent = (
				<>
					due in{' '}
					<span className="due-soon">{humanizedDaysUntilDue}</span>
				</>
			);
		} else if (daysUntilDue > 0) {
			dueComponent = (
				<>
					due{' '}
					<span className="due-very-soon">
						{daysUntilDue === 1 ? 'tomorrow' : 'today'}
					</span>
				</>
			);
		} else {
			dueComponent = (
				<span className="due-past">
					{humanizedDaysUntilDue} past due
				</span>
			);
		}
	}

	return (
		<div style={{ display: 'flex', flexDirection: 'column', flexGrow: 1 }}>
			<div className="note-card-description card-content-text">
				{description}
			</div>
			{dueComponent && <div className="due">{dueComponent}</div>}
		</div>
	);
};

const NoteCard: React.FC<NoteCardProps> = ({ note }) => (
	<Card
		title={note.name}
		content={
			<NoteCardContent
				description={note.description}
				startDate={note.startDate}
				dueDate={note.dueDate}
			/>
		}
		isImageDisabled
		size={'sm'}
		className="note"
	/>
);

export default NoteCard;
