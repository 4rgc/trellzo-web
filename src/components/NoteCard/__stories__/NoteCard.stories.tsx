import { ComponentMeta, ComponentStory } from '@storybook/react';
import { DateTime, Duration } from 'luxon';
import NoteCard from '../NoteCard';

export default {
	title: 'Components/NoteCard',
	args: {
		note: {
			_id: '1',
			name: 'Buy Groceries',
			description: 'eggs, chicken, tomatoes, green onion',
		},
	},
} as ComponentMeta<typeof NoteCard>;

const Template: ComponentStory<typeof NoteCard> = (args) => (
	<div style={{ width: '200px' }}>
		<NoteCard {...args} />
	</div>
);

export const Default = Template.bind({});

export const LongName = Template.bind({});
LongName.args = {
	note: {
		_id: '1',
		name: 'hello my name is asdf nice to meet you',
		description: 'description',
	},
};

export const LongDesctription = Template.bind({});
LongDesctription.args = {
	note: {
		_id: '1',
		name: 'Name is short',
		description:
			'description is very very very very very very very very very very very very long',
	},
};

export const DueIn5Days = Template.bind({});
DueIn5Days.args = {
	note: {
		_id: '1',
		name: 'Name',
		description: 'Description',
		dueDate: DateTime.now()
			.plus(Duration.fromDurationLike({ days: 5 }))
			.toISO(),
	},
};

export const DueIn3Days = Template.bind({});
DueIn3Days.args = {
	note: {
		_id: '1',
		name: 'Name',
		description: 'Description',
		dueDate: DateTime.now()
			.plus(Duration.fromDurationLike({ days: 3 }))
			.toISO(),
	},
};

export const DueTomorrow = Template.bind({});
DueTomorrow.args = {
	note: {
		_id: '1',
		name: 'Name',
		description: 'Description',
		dueDate: DateTime.now()
			.plus(Duration.fromDurationLike({ days: 1 }))
			.toISO(),
	},
};

export const DueToday = Template.bind({});
DueToday.args = {
	note: {
		_id: '1',
		name: 'Name',
		description: 'Description',
		dueDate: DateTime.now().toISO(),
	},
};

export const OneDayPastDue = Template.bind({});
OneDayPastDue.args = {
	note: {
		_id: '1',
		name: 'Name',
		description: 'Description',
		dueDate: DateTime.now().minus({ days: 1 }).toISO(),
	},
};
