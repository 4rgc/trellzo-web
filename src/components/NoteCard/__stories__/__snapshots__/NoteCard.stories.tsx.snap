// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/NoteCard Default smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":r0:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":r0:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:r0:-hidden-text-:r1:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":r0:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        Buy Groceries
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          eggs, chicken, tomatoes, green onion
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/NoteCard DueIn3Days smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":rg:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":rg:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:rg:-hidden-text-:rh:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":rg:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        Name
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          Description
        </div>
        <div class="due">
          due in
          <span class="due-soon">
            3 days
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/NoteCard DueIn5Days smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":rc:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":rc:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:rc:-hidden-text-:rd:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":rc:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        Name
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          Description
        </div>
        <div class="due">
          due in
          <span class="due-far">
            5 days
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/NoteCard DueToday smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":ro:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":ro:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:ro:-hidden-text-:rp:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":ro:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        Name
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          Description
        </div>
        <div class="due">
          due
          <span class="due-very-soon">
            today
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/NoteCard DueTomorrow smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":rk:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":rk:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:rk:-hidden-text-:rl:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":rk:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        Name
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          Description
        </div>
        <div class="due">
          due
          <span class="due-very-soon">
            tomorrow
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/NoteCard LongDesctription smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":r8:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":r8:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:r8:-hidden-text-:r9:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":r8:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        Name is short
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          description is very very very very very very very very very very very very long
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/NoteCard LongName smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":r4:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":r4:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:r4:-hidden-text-:r5:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":r4:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        hello my name is asdf nice to meet you
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          description
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/NoteCard OneDayPastDue smoke-test 1`] = `
<div data-rfd-droppable-id="1"
     data-rfd-droppable-context-id=":rs:"
     style="width: 200px;"
>
  <div class="card card-sm card-noimage note"
       data-rfd-draggable-context-id=":rs:"
       data-rfd-draggable-id="1"
       tabindex="0"
       role="button"
       aria-describedby="rfd-hidden-text-:rs:-hidden-text-:rt:"
       data-rfd-drag-handle-draggable-id="1"
       data-rfd-drag-handle-context-id=":rs:"
       draggable="false"
  >
    <div class="card-container-no-text">
      <h3 class="card-title">
        Name
      </h3>
      <div style="display: flex; flex-direction: column; flex-grow: 1;">
        <div class="note-card-description card-content-text">
          Description
        </div>
        <div class="due">
          <span class="due-past">
            1 day past due
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
